src/org/nfrac/comportex/cells.cljc

```diff
@@ -370,16 +371,23 @@
   the proximal excitation but is given keyed by cell rather than by
   column. Normally only cells with some proximal input are included,
   but if `spontaneous-activation?` is true, this is relaxed
-  (i.e. prediction alone could cause activation)."
+  (i.e. prediction alone could cause activation).
+
+  * col-exc is keyed by column as [col 0].
+  * tp-exc is keyed by cell as [col ci]."
   [col-exc tp-exc distal-exc distal-weight spontaneous-activation? depth]
-  (let [;; add TP columns to list of columns to consider (may have no proximal)
-        basic-col-exc (merge-with + col-exc
-                                  (into {} (map (fn [[col _]] [col 0])) (keys tp-exc)))
+  (let [has-tp? (seq tp-exc)
+        ;; add TP columns to list of columns to consider (may have no proximal)
+        basic-col-exc (if has-tp?
+                        (merge-with + col-exc
+                                    (into {} (map (fn [[col _]] [[col 0] 0]))
+                                          (keys tp-exc)))
+                        col-exc)
         ;; expand to all cells within columns, add TP values
-        basic-exc (for [[col exc] basic-col-exc
+        basic-exc (for [[[col _] exc] basic-col-exc
                         ci (range depth)
                         :let [cell-id [col ci]
-                              tp (get tp-exc cell-id 0.0)]]
+                              tp (if has-tp? (get tp-exc cell-id 0.0) 0.0)]]
                     [cell-id (+ exc tp)])]
     (if (zero? distal-weight)
       (into {} basic-exc)
```

Comment:
Another "not specific to this commit": A couple lines below this, it has:

~~~clojure
(merge (zipmap (keys distal-exc) (repeat 0.0))
       (into {} basic-exc))
~~~

This would be faster as:

~~~clojure
(into (zipmap (keys distal-exc) (repeat 0.0))
      basic-exc)
~~~